### ISSUE TITLE
Renaming a node loses nodepath references from that node to children (and vice versa), and references within the branch.

### ISSUE THREAD

**Issue by takaturre** [godotengine/godot#35226]
**Godot version:**
3.2 beta5 or any earlier I've tried.

**Issue description:**
When renaming a node that has node path references to children nodes (any nested inside), those references are kept after first rename, but renaming it again loses the references - they are left pointing at the situation after the first renaming. In addition to the renamed node losing its references, it also happens for the kids: any nested child node referring to the renamed node (by nodepath) loses the tie - exactly in the same way as above (after 1st works, after 2nd broken). 
**Edit:** #30534 seems very related. Note also that the situation "breaks" (internally?) already after the first renaming. Before that it works, after that, no more renaming support.

**Steps to reproduce:**
- Create a node with a script that has an export variable of NodePath type (or use an internal node that uses those): `export(NodePath) var my_path`
- Duplicate that node and move it inside the original node - now we have two nodes nested.
- Assign the parent node's `my_path` to the kid, and the kid's `my_path` to the parent.
- Rename the parent node (eg. to `name_after_1st`) and check the situation: Seems refs are correct (by inspector UI look as well as hovering to see the path, as well as trying by script).
- Rename again (eg. to `name_after_2nd`) and check the situation: Now the parent refers to `../name_after_1st/kid` where as the kid refers to `../../name_after_1st`. So both refs are broken.

**Workarounds:**
- I currently have a system that tracks those special nodes (using node path refs) using mostly `_enter_tree` and `_exit_tree` functionality (+ some idle_frame waits and switching scene tabs special hacks). The core of the solution is to split the path into two: path to the common parent and then the sub path from it to the target node. Although this is hacky and annoying and adds boilerplate to each script using node path refs in editor (there are hundreds of them), it works for all cases except renaming the root node - which I haven't found a solution for. A tiny bit related is #35222 (problems on changing type of the root node). 

**Notes on the above workaround for renaming and moving (copied from the script):**

Background:
- When a node is moved, Godot will automatically update its new location to any node in the scene that refers to it (by NodePath).
- Problem is that, if the updated node is moving as well, the new location will be wrong (it should not be updated at all).

Solution:
1. We can use the ancestor part of the updated node path to get the common parent. Eg. in "../../../something/new_location", the "../../../" gives us the common parent.
2. We split the path at the common ancestor. Upon exiting we store the common ancestor as a node + then the sub path down from it.
3. Upon entering, we get the stored common parent and sub path, get the node at the sub path, and update the NodePath ref by getting path to that sub node.

**Update:**
The scope of the issue seems to be larger: Renaming a node, will break all node path references that point **from within** the node **to within** the node. So in the scenario below, `NodeWithNodePath` has a export(NodePath) variable reference to `TargetNode`, and renaming the container `NodeToRename` will break this path - in the way described above.
```
> NodeToRename
> NodeToRename > NodeWithNodePath
> NodeToRename > TargetNode
```
Potential solution:
If Godot really stores the paths as relative, then the solution would be to simply skip the renaming procedure for nodes that are within the renamed node (children or grand-children) - as they were correct, no smart-renaming needed.

**Comment by bojidar-bg** [comment]
Potentially related to #34872, can't say for sure.

Does the issue happen if:
1. You save the scene before the second rename?
2. You save and reload (close and open again) the scene before the second rename?
3. You save and reload the scene before the first rename?

**Comment by takaturre** [comment]
1. Yes, the issue still happens. 
2. Yes, exactly like 1.
3. Yes.
In short, seems completely unrelated to saving. Due to this, I had earlier thought the first rename already broke it, but in the message above I thought it was a clumsy memory (because thought that if the script gets the value after first rename, it should work).

**Comment by takaturre** [comment]
So.. My last completely uneducated guess is that Godot internally stores those nodes (not node paths) for the renaming conveniencies. But after the first renaming, this storage is lost (due to renaming), and so the "renaming support" vanishes afterwards. (I have 0 knowledge of Godot's internal workings.)

**Comment by tom-jk** [comment]
Hello, I got bit by this one yesterday (v3.2.1.stable.official). If it's ok, I'd like to add a simple example that people could test against.

[NodepathRenameBugTest.zip](https://github.com/godotengine/godot/files/4563939/NodepathRenameBugTest.zip)

1. Open the project, Main.tscn should be open with four instances of Character.tscn. Run the game, all instances animate.
2. Select an instance node. Use the Instance a Scene File button and add a new instance of Character.tscn. It will be added under the selected node and be called “Character”.
3. Drag and drop it onto the Main node in the tree. It renames itself to “Character5”.
4. Open Character.tscn. Run the game again. Now only the “Character” instance animates. Look in the Character.tscn AnimationPlayer; if you double click the track name (“position”) you’ll see its full path is “../Character/Sprite:position”, when it should be “Sprite:position”.
5. Go back into Main.tscn and delete the “Character” node. Look again in Character.tscn and the animation is now empty.
6. Close Godot and edit Character.tscn to remove the “../Character/” prefix on line 11 to solve the problem. Or, undo to before deleting the node in step 5, double click the track name and edit out the prefix there.

To help searchers: this bug might show up in your game as all characters of a type no longer animating, except for one instance that's freaking out because it's receiving all of their animation play instructions.

**Comment by noidexe** [comment]
I just reproduced this together with my every one of my 30 pupils 🤦‍♂️ so I can confirm it's still happening on 3.2.3
Thanks @tom-jk for the repro steps. 
Is this marked as a known-issue? It's pretty critical since it's really hard to debug if you don't know what's going on or that it is breaking other scenes.

**Comment by rcorre** [comment]
Looks like there are two PRs open: https://github.com/godotengine/godot/pull/35319 and https://github.com/godotengine/godot/pull/42314

**Comment by noidexe** [comment]
This seems to have resurfaced in 3.3.4
![image](https://user-images.githubusercontent.com/526829/136500602-8cccc66b-57eb-48d7-bb8a-e8386ae2157c.png)

Repro steps:
- instance a Character.tscn as a child of the instance named "Character"
- rename the newly instance node: "Character" -> "Character4"
- drag the renamed node on top of the root node ("Main")
- the node will rename itself "Character5"
- Open Character.tscn and check the AnimationPlayer. The track is now ../Character4/Sprite:position

**Comment by mbrlabs** [comment]
@noidexe can you reproduce this in 3.3 (first version which included my fix)? I did some basic testing without the AnimationPlayer in 3.3.4 (just renaming and moving plain old nodes with exported NodePaths) and that still works. This might be an issue in AnimationPlayer. 

I also remember that @pouleyKetchoupp did some more fixes to NodePaths for the 3.4 release. Maybe it's fixed in that version.

**Comment by akien-mga** [comment]
Thanks for testing! Closing as fixed in 3.4 then, which will soon be released.
